src/include/miopen/tensor_layout.hpp

```diff
@@ -79,6 +89,123 @@ inline std::string tensor_layout_get_default(int size)
     return "";
 }
 
+template <class Vector, class Op>
+static inline std::vector<int64_t> sort_permutation(const Vector& data, Op op)
+{
+    std::vector<std::int64_t> result(data.size());
+    std::iota(result.begin(), result.end(), 0);
+    std::sort(result.begin(), result.end(), [&](auto x, auto y) { return op(data[x], data[y]); });
+    return result;
+}
+
+static void construct_all_layouts(std::vector<std::string>& all_layouts,
+                                  std::string tmp,
+                                  const std::string& labels,
+                                  std::vector<bool> is_recurrent)
+{
+    if(tmp.size() == labels.size())
+    {
+        // std::cout << tmp << std::endl;
+        all_layouts.push_back(tmp);
+        return;
+    }
+
+    for(int i = 0; i < labels.size(); ++i)
+    {
+        // Skip current label if it already exists
+        if(find(tmp.begin(), tmp.end(), labels[i]) != tmp.end())
+            continue;
```

Comment:
TODO: use an additional `is_used` vector instead of find to reduce time complexity.